### PR TITLE
SAMZA-1247: MessageStreamImpl#merge shouldn't mutate input collection

### PR DESCRIPTION
**Pull request by prateekm**
Also fixes
SAMZA-1253: MessageStream.merge operator broken for nested types

### CODE REVIEW

**Comment by prateekm** [comment]
@xinyuiscool @vjagadish1989 please take a look.

**Review by vjagadish1989 — APPROVED** [review]
looks good to me. approved. thank you for the contribution Prateek.

File: samza-core/src/main/java/org/apache/samza/operators/MessageStreamImpl.java [hunk 1]
```diff
@@ -183,14 +185,15 @@ public void init(Config config, TaskContext taskContext) {
...
 
   @Override
-  public MessageStream<M> merge(Collection<MessageStream<? extends M>> otherStreams) {
+  public MessageStream<M> merge(Collection<? extends MessageStream<? extends M>> otherStreams) {
     MessageStreamImpl<M> nextStream = new MessageStreamImpl<>(this.graph);
-
-    otherStreams.add(this);
-    otherStreams.forEach(other -> {
-        OperatorSpec mergeOperatorSepc =
+    List<MessageStream<M>> streamsToMerge = new ArrayList<>();
```
Comment (jmakes) [inline samza-core/src/main/java/org/apache/samza/operators/MessageStreamImpl.java]: Why not leverage the ArrayList constructor  (e.g. new ArrayList<>(otherStreams)) to copy the otherStreams collection and then add "this"?

**Comment by prateekm** [comment]
@jmakes updated with feedback.

**Pull request closed without merging**